### ISSUE TITLE
test failure with R ≥ 3.6: 'The R package "stats" is not installed'

### ISSUE THREAD

**Issue by pauleve** [rpy2/rpy2#720]
I'm trying to upgrade the conda-forge package to 3.3.5, but the validation failed during the testsuite (`rpy2/tests/robjects/test_packages.py`), throwing an error
```
        if not isinstalled(name, lib_loc=lib_loc):
            raise PackageNotInstalledError(
>               'The R package "%s" is not installed.' % name
            )
E           rpy2.robjects.packages.PackageNotInstalledError: The R package "stats" is not installed.
```
Full log details as https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=184848&view=logs&j=4f922444-fdfe-5dcf-b824-02f86439ef14&t=b2a8456a-fb11-5506-ca32-5ccd32538dc0&l=2447

I'm not sure where to go from there, as r-base 3.6.3 is installed... any hint?

The conda recipe is here: https://github.com/pauleve/rpy2-feedstock/tree/3.3.5_hbe2245/recipe

**Comment by lgautier** [comment]
Hard you running unit tests in parallel by any chance ?

**Comment by pauleve** [comment]
No, the testsuit command is `pytest -s -rxs -v rpy2/tests -k "not test_display_args"`

**Comment by lgautier** [comment]
The R package `stats` is part of R's "stdlib". Errors when trying to load it are usually symptomatic of an embedded R in jeopardy (and there are many ways to reach that state when dealing with the R C API).

Can you run the following debug help in your CI script ?
(you can run it either from a git clone, a source directory, or with an already installed `rpy2` package.)

```bash
python -m rpy2.situation
```

This will give you information about the R found by `rpy2`.

**Comment by xhochy** [comment]
Debugging @pauleve 's problem, for me the output of `python -m rpy2.situation` is 

```
+ python -m rpy2.situation
rpy2 version:
3.3.5
Python version:
3.8.5 | packaged by conda-forge | (default, Jul 22 2020, 17:24:51)
[Clang 10.0.0 ]
Looking for R's HOME:
    Environment variable R_HOME: None
    Calling `R RHOME`: $PREFIX/lib/R
    Environment variable R_LIBS_USER: None
R's additions to LD_LIBRARY_PATH:

R version:
    In the PATH: R version 3.6.3 (2020-02-29) -- "Holding the Windsock"
    Loading R library from rpy2: OK
Additional directories to load R packages from:
None
C extension compilation:
  include:
  ['$PREFIX/lib/R/include']
  libraries:
  ['R', 'pcre', 'lzma', 'bz2', 'z', 'dl', 'm', 'iconv', 'icuuc', 'icui18n']
  library_dirs:
  ['$PREFIX/lib', '$PREFIX/lib/R/lib']
  extra_compile_args:
  []
  extra_link_args:
  ['-fopenmp', '-Wl,-dead_strip_dylibs', '-Wl,-pie', '-Wl,-headerpad_max_install_names', '-Wl,-dead_strip_dylibs', '-Wl,-rpath,$PREFIX/lib']
```

**Comment by xhochy** [comment]
Is this maybe an issue with the test? Looking into this with the debugger, it seems like `lib_loc` is incorrect:

```
> /Users/uwe/miniconda3/conda-bld/rpy2_1595498961554/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/python3.8/site-packages/rpy2/robjects/packages.py(470)importr()
-> raise PackageNotInstalledError(
(Pdb) p lib_loc
'/Users/uwe/miniconda3/conda-bld/rpy2_1595498961554/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/R/library/stats'
(Pdb) p isinstalled(name)
True
(Pdb) p isinstalled(name, lib_loc=lib_loc)
False
(Pdb) p isinstalled(name, lib_loc='/Users/uwe/miniconda3/conda-bld/rpy2_1595498961554/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/R/library/stats')
False
(Pdb) p isinstalled(name, lib_loc='/Users/uwe/miniconda3/conda-bld/rpy2_1595498961554/_test_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/R/library')
True
```

**Comment by kiwifb** [comment]
I am having the exact same issue in Gentoo (r 3.6.3, rpy 3.3.5)
```
____________________________________________________________________ TestImportr.test_import_stats_with_libloc ____________________________________________________________________

self = <rpy2.tests.robjects.test_packages.TestImportr object at 0x7fe9cc30bb50>

    def test_import_stats_with_libloc(self):
        path = robjects.packages_utils.get_packagepath('stats')
        stats = robjects.packages.importr('stats',
                                          on_conflict='warn',
>                                         lib_loc = path)

rpy2/tests/robjects/test_packages.py:102: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

name = 'stats', lib_loc = '/usr/lib64/R/library/stats', robject_translations = {}, signature_translation = True, suppress_messages = True, on_conflict = 'warn'
symbol_r2python = <function default_symbol_r2python at 0x7fe9d0084170>, symbol_resolve = <function default_symbol_resolve at 0x7fe9d0084200>, data = True

    def importr(name,
                lib_loc=None,
                robject_translations={},
                signature_translation=True,
                suppress_messages=True,
                on_conflict='fail',
                symbol_r2python=default_symbol_r2python,
                symbol_resolve=default_symbol_resolve,
                data=True):
        """ Import an R package.
    
        Arguments:
    
        - name: name of the R package
    
        - lib_loc: specific location for the R library (default: None)
    
        - robject_translations: dict (default: {})
    
        - signature_translation: (True or False)
    
        - suppress_message: Suppress messages R usually writes on the console
          (defaut: True)
    
        - on_conflict: 'fail' or 'warn' (default: 'fail')
    
        - symbol_r2python: function to translate R symbols into Python symbols
    
        - symbol_resolve: function to check the Python symbol obtained
                          from `symbol_r2python`.
    
        - data: embed a PackageData objects under the attribute
          name __rdata__ (default: True)
    
        Return:
    
        - an instance of class SignatureTranslatedPackage, or of class Package
    
        """
    
        if not isinstalled(name, lib_loc=lib_loc):
            raise PackageNotInstalledError(
>               'The R package "%s" is not installed.' % name
            )
E           rpy2.robjects.packages.PackageNotInstalledError: The R package "stats" is not installed.

rpy2/robjects/packages.py:471: PackageNotInstalledError
```
and I can't help feeling that indeed `lib_loc = '/usr/lib64/R/library/stats` is wrong. It feels one level too deep to be a library path for R packages. Which means that either `robjects.packages_utils.get_packagepath` doesn't do what it is supposed to do, or the person writing the test doesn't. Either way, I concerned that if the test pass on any platforms.

**Comment by kiwifb** [comment]
For the record I am not using conda, so the tag feels misleading.

**Comment by lgautier** [comment]
I just added that tag and looked at the initial issue report. This is reported on conda, there has been conda-specific issues with similar symptoms over the years, and this is otherwise not widely reported for what is otherwise an error on import.

It does not imply that it is impossible to reproduce what conda is doing to obtain the same error.

Otherwise, a path for R libraries that is like `${R_HOME}/library/stats` is the norm in my experience. Continuous automated testing is performed on Linux (Ubuntu), OSX, and Windows, each with several version of Python and tests are passing (although I can occasionally break them temporarily).

**Comment by kiwifb** [comment]
I would agree the path to the package is correct. But the path to the package itself isn't a library location, the package is inside the library (location). So, it feels weird to me that this test would pass as, is unless something about library location is slightly different from version to version or configuration. In any case it should be identified.

I am trying to update the rpy package in Gentoo, so I am using all proper system installation and that surfaced as a test failure (along with a few other bits that are probably due to newer pexpect and numpy). And unless library location is not working like I expect it to be, I don't see how this particular test, and the next one are supposed to pass.

**Comment by lgautier** [comment]
I see what you mean. I scanned through the issue a little too quickly.

That part of the code was touched relatively recently (de86402c6a3b3c3ec2946ee3638ec1f2dc743871), but if it broke something I wonder why it didn't break the CI.

**Comment by kiwifb** [comment]
Could it be that R-4 changed the behavior slightly? In that case it wouldn't show up if the CI focuses on the latest R.

**Comment by lgautier** [comment]
This would be an unfortunate coincidence, and it would seem quite surprising (the path for R packages has not changed since... forever).

I am looking at what is happening...

**Comment by lgautier** [comment]
I can reproduce with R-4.0. Looks like I have a few things to chase now .

**Comment by lgautier** [comment]
The unit test was wrong. The fix will be in rpy2-3.3.6.

Now I have to look at why the CI did not break.

**Comment by kiwifb** [comment]
Let me know what you find. I am "professionally" interested in why those things happen.

**Comment by lgautier** [comment]
Murphy's Law. The script to run tests was not stopping at the first testing layer (the fix is trivial: 0de3965ac6019f5fa0c2ddffc048c094aac746eb). I usually run tests locally before pushing but it looks I forgot for that recent commit.
